File: 2023/25.rb
input = File.read(File.basename(__FILE__).gsub('rb', 'input')).strip.split("\n")

graph = {}
input.each do |line|
  left, right = line.split(': ')
  right.split(' ').each do |r|
    graph[left] ||= []
    graph[left] << r
    graph[r] ||= []
    graph[r] << left
  end
end

connections = {}
graph.keys.each do |node|
  visited = {}

  queue = [node]
  until queue.empty?
    current = queue.shift
    visited[current] = true

    graph[current].each do |to|
      next if visited[to] || queue.include?(to)

      connections[[current, to].sort] ||= 0
      connections[[current, to].sort] += 1

      queue << to
    end
  end
end

connections.sort_by { _2 }.last(3).map { _1[0] }.each do |node1, node2|
  graph[node1].delete(node2)
  graph[node2].delete(node1)
end

visited = {}
queue = [graph.keys.first]

until queue.empty?
  current = queue.shift
  visited[current] = true

  graph[current].each do |to|
    next if visited[to] || queue.include?(to)

    queue << to
  end
end

p (graph.size - visited.size) * visited.size

File: 2023/25.py
from collections import defaultdict

input = open(f"{__file__.replace('py', 'input')}").read().strip().split("\n")

graph = defaultdict(lambda: [])
for line in input:
    left, right = line.split(": ")
    for r in right.split(" "):
        graph[left].append(r)
        graph[r].append(left)

connections = defaultdict(lambda: 0)
for node in graph.keys():
    visited = {}

    queue = [node]
    while queue:
        current = queue.pop(0)
        visited[current] = True

        for to in graph[current]:
            if to in visited or to in queue:
                continue

            connections[":".join(sorted([current, to]))] += 1

            queue.append(to)

for nodes, _ in sorted(connections.items(), key=lambda x: x[1])[-3:]:
    node1, node2 = nodes.split(":")
    graph[node1].remove(node2)
    graph[node2].remove(node1)

visited = {}
queue = [list(graph.keys())[0]]

while queue:
    current = queue.pop(0)
    visited[current] = True

    for to in graph[current]:
        if to in visited or to in queue:
            continue

        queue.append(to)

print((len(graph) - len(visited)) * len(visited))
